File: src/main/java/org/hisp/dhis/lib/expression/syntax/FragmentContext.java
package org.hisp.dhis.lib.expression.syntax;

import java.util.function.Function;

public interface FragmentContext {

        /*
        Lookup named building blocks (non-terminals)
    */

    Fragment fragment(String name);

    static Fragment lookup(Expr expr, Function<Expr, String> parseName, Function<String, Fragment> lookup) {
        int s = expr.position();
        String name = parseName.apply(expr);
        Fragment res = lookup.apply(name);
        if (res == null)
        {
            expr.error(s, "Unknown function or constant: '"+name+"'");
        }
        return res;
    }
}

File: src/jvmMain/java/org/hisp/dhis/lib/expression/syntax/FragmentContext.kt
package org.hisp.dhis.lib.expression.syntax

import java.util.function.Function

fun interface FragmentContext {
    /*
        Lookup named building blocks (non-terminals)
    */
    fun fragment(name: String): Fragment?

    companion object {
        fun lookup(expr: Expr, parseName: Function<Expr, String>, lookup: Function<String, Fragment?>): Fragment {
            val s = expr.position()
            val name = parseName.apply(expr)
            val res = lookup.apply(name)
            if (res == null) {
                expr.error(s, "Unknown function or constant: '$name'")
            }
            return res!!
        }
    }
}
